third_party/move/move-compiler-v2/src/env_pipeline/model_ast_lints.rs

```diff
@@ -31,24 +35,40 @@ pub fn checker(env: &mut GlobalEnv) {
 /// expression as we traverse the model AST.
 /// Implement at least one of the `visit` methods to be a useful lint.
 trait ExpressionLinter {
-    /// The name of the lint.
-    fn get_name(&self) -> &'static str;
+    /// The corresponding lint checker enumerated value.
+    fn get_lint_checker(&self) -> LintChecker;
 
     /// Examine `expr` before any of its children have been visited.
-    /// Potentially emit lint warnings using `env.lint_diag()`.
+    /// Potentially emit lint warnings using `self.warning()`.
     fn visit_expr_pre(&mut self, _env: &GlobalEnv, _expr: &ExpData) {}
 
     /// Examine `expr` after all its children have been visited.
-    /// Potentially emit lint warnings using `env.lint_diag()`.
+    /// Potentially emit lint warnings using `self.warning()`.
     fn visit_expr_post(&mut self, _env: &GlobalEnv, _expr: &ExpData) {}
+
+    /// Emit a lint warning with the `msg` highlighting the `loc`.
+    fn warning(&self, env: &GlobalEnv, loc: &Loc, msg: &str) {
```

Comment:
While it can redefined by the lint checker on a need by need basis, the expectation is that most linters will just use this as is.

This method dispatches back `self.get_lint_checker()` to get the lint name from the checker, so this default implementation makes use of other trait methods (I saw a similar pattern in the Rust book: https://doc.rust-lang.org/book/ch10-02-traits.html#default-implementations, and found it to be reasonable).